### ISSUE TITLE
Bug: Duplicate JPEGs are not being removed

### ISSUE THREAD

**Issue by 2320sharon** [Doodleverse/CoastSeg#196]
### Description:
Users have found that duplicate images are being downloaded and used to extract shorelines. Downloading duplicate images has always been an issue with coastsat's download workflow, but the question is the process of removing the duplicates only happening to the tifs and not removing duplicate jpegs? Users are also wondering if its possible to modify the download workflow so that duplicates are detected before they are downloaded so that downloading duplicate images does not further slow down the downloads. Issues with duplicates are most prevalent with S2 imagery. This has led to significant delays in download times, impacting user experience and overall workflow efficiency.

#### Concerns:
- The removal of duplicate images seems to be limited to TIFFs only, leaving behind duplicate JPEGs.
- Users have expressed a need for refining the download workflow to identify and exclude duplicate images before the download process begins, ideally improving workflow speed.

### Tasks:
- [ ] **Task 1:** Investigate whether the “remove duplicates” function is currently excluding JPEGs and ensure all duplicate JPEGs are removed in the process.
- [ ] **Task 2:** Validate the sequence of the workflow, ensuring the “remove duplicates” function is executed before the shoreline extraction begins.
- [ ] **Task 3:** Explore the feasibility of identifying duplicate images in the dataset prior to initiating downloads from Google Earth Engine, possibly through advanced filtering techniques.

### Acceptance Criteria:
- The “remove duplicates” function should effectively remove both TIFF and JPEG duplicates.

**Comment by 2320sharon** [comment]
## Relevant Research Findings
- The `remove_duplicates` defined in coastsat removes duplicate shorelines from the Shoreline dictionary. It does not remove duplicate tiffs or jpegs
- duplicate tips are typically S2 imagery
- `handle_duplicate_image_names` defined in coastsat  renames the duplicate tiffs by adding `dup_X` to the file name. These duplicate files are images from the same satellite collection with the same image timestamp. These files are temporary tiffs that are later converted into the real tiffs that are saved.
- the filtering of the S2 collection begins before the download ever starts. 
`im_dict_T1["S2"] = filter_S2_collection(im_dict_T1["S2"])`
Basically the `filter_S2_collection` function deletes all the S2 imagery with the same time stamp and different UTM zones. Images with the same timestamp and the same UTM zone keep only the first one.

**Comment by 2320sharon** [comment]
These two jpegs  `2018-12-03-18-39-48_RGB_L8.jpg`  and `2018-12-03-18-40-12_RGB_L8.jpg` were captured on the same day
`2018-12-03` and at almost the same time `18-39-48` and `18-40-12` would these images be considered duplicates since they are less than 24 hours apart? @dbuscombe-usgs 
![2018-12-03-18-39-48_RGB_L8](https://github.com/Doodleverse/CoastSeg/assets/61564689/32015fd7-a90d-479d-a1f9-1ac5846038fd7)
![2018-12-03-18-40-12_RGB_L8](https://github.com/Doodleverse/CoastSeg/assets/61564689/6d616322-d983-49c9-abe5-e1a76d13920c)

**Comment by dbuscombe-usgs** [comment]
We'd want to keep both images. That's super valuable having images on consecutive days! 

Duplicates are only when images are identical times

**Comment by 2320sharon** [comment]
> We'd want to keep both images. That's super valuable having images on consecutive days!
> 
> Duplicates are only when images are identical times

Ah good to know, thanks for helping me double check that.

I have validated that even though duplicated S2 imagery are supposed to be filtered out by the `filter_S2_collection` function duplicates are still getting through. When I ran the script below on 700 S2 images I downloaded I got the outputs below. I did not show all of the duplicates because there are a lot of them.

```python
import os 
from collections import defaultdict

file_list=os.listdir('/home/sha23/development/coastseg/CoastSeg/data/ID_kyg2_datetime10-02-23__03_11_52/jpg_files/preprocessed/RGB')
grouped_files = defaultdict(list)
# Extract date and group files
for filename in file_list:
    date_part = filename.split('_')[0]
    date_only = '-'.join(date_part.split('-')[:3])  # extract the date
    grouped_files[date_only].append(filename)

# Print the result
for date, files in grouped_files.items():
    if len(files)>2:
        print(date)
        for file in files:
            print(f"  {file}")
        print()
```
### Output
2019-08-04
  2019-08-04-19-23-30_RGB_S2.jpg
  2019-08-04-19-23-30_RGB_S2.jpg
  2019-08-04-19-23-30_RGB_S2.jpg

2021-10-14
  2021-10-14-19-13-31_RGB_S2.jpg
  2021-10-14-19-13-31_RGB_S2.jpg
  2021-10-14-19-13-31_RGB_S2.jpg

2022-02-19
  2022-02-19-19-23-27_RGB_S2.jpg
  2022-02-19-19-23-27_RGB_S2.jpg
  2022-02-19-19-23-27_RGB_S2.jpg

2018-03-22
  2018-03-22-19-21-33_RGB_S2.jpg
  2018-03-22-19-21-33_RGB_S2.jpg
  2018-03-22-19-21-33_RGB_S2.jpg

2019-02-27
  2019-02-27-19-13-24_RGB_S2.jpg
  2019-02-27-19-13-24_RGB_S2.jpg
  2019-02-27-19-13-24_RGB_S2.jpg

2018-03-19
  2018-03-19-19-07-40_RGB_S2.jpg
  2018-03-19-19-07-40_RGB_S2.jpg
  2018-03-19-19-07-40_RGB_S2.jpg

2019-05-01
  2019-05-01-19-23-33_RGB_S2.jpg
  2019-05-01-19-23-33_RGB_S2.jpg
  2019-05-01-19-23-33_RGB_S2.jpg

2022-03-13
  2022-03-13-19-13-27_RGB_S2.jpg
  2022-03-13-19-13-27_RGB_S2.jpg
  2022-03-13-19-13-27_RGB_S2.jpg

2021-06-19
  2021-06-19-19-23-26_RGB_S2.jpg
  2021-06-19-19-23-26_RGB_S2.jpg
  2021-06-19-19-23-26_RGB_S2.jpg

2021-07-26
  2021-07-26-19-13-31_RGB_S2.jpg
  2021-07-26-19-13-31_RGB_S2.jpg
  2021-07-26-19-13-31_RGB_S2.jpg

**Comment by 2320sharon** [comment]
I ran this script across all the data I've downloaded and I didn't find any duplicates
```
import os 
from collections import defaultdict
from collections import Counter

data_directory = r'C:\development\doodleverse\coastseg\CoastSeg\data'
roi_dirs =  os.listdir(data_directory)
for roi_dir in roi_dirs:
    jpeg_directory =  os.path.join(roi_dir,"jpg_files","preprocessed","RGB")
    if os.path.exists(jpeg_directory):
        file_list=os.listdir(jpeg_directory)
        counter = Counter(file_list)
        duplicates = {file: count for file, count in counter.items() if count > 1}

        # Print the duplicates
        for duplicate, count in duplicates.items():
            print(f"Filename: {duplicate} - Count: {count}")
```

**Comment by 2320sharon** [comment]
@dbuscombe-usgs have you found duplicate imagery in any of the downloads you've performed?

**Comment by 2320sharon** [comment]
I heard back from Catherine on the duplicate images issue and here is what she said:

> I'm going through the images currently. It appears that the images may actually have unique IDs, but they were so extremely similar with the exact date, hour, and even minutes for some. There are some days with S2 that have 2-3 images for the same day and extremely similar times, hence I thought it was identical. Here is an example that is only 15 mins away from each other. I haven't seen it with Landsat yet. This happens with nearly 75% of S2 images after 2018.

![2022-11-18-15-52-02_RGB_S2](https://github.com/Doodleverse/CoastSeg/assets/61564689/3c93c804-6423-4a84-a376-33e9f6bae346)
![2022-11-18-15-51-47_RGB_S2](https://github.com/Doodleverse/CoastSeg/assets/61564689/d54a6642-43c7-44e0-8ae6-4660136b080b)

So it seems there aren't identical images being generated just multiple images that are sometimes seconds apart.
@dbuscombe-usgs do we want to keep these images that are minutes/seconds apart?

**Comment by 2320sharon** [comment]
During the meeting today we addressed the confusion about "duplicates" images or more accurately images  captured within a few minutes of each other or less. We determined that it would be easiest to make this a post-processing script that removes images that are less than a few minutes apart. It was suggested that this would be a script located in the scripts directory.

**Comment by 2320sharon** [comment]
- [ ] write a script the removes all images within a designated time frame from other imagery.
@dbuscombe-usgs maybe images within 5-10 minutes of each other should be removed?

**Comment by dbuscombe-usgs** [comment]
I think the point of the script would be for the user to specify what time period that like, and it should go in SDS-tools
